### PR TITLE
Migrate core to Typescript

### PR DESCRIPTION
**Pull request by Etheryte**
## What does this PR change?

~~Followup to https://github.com/uyuni-project/uyuni/pull/3106, no need to review this until that gets merged.~~

Migrate all code in `web/html/src/core` to Typescript.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- No tests: Needs to pass all existing tests.

- [x] **DONE**

## Links

Related: https://github.com/SUSE/spacewalk/issues/13208

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: web/html/src/core/spa/spa-renderer.ts [hunk 1]
```diff
@@ -1,87 +1,83 @@
...
 }
 
 function beforeNavigation() {
-  window.pageRenderers.spa.previousReactRenderers = window.pageRenderers.spa.reactRenderers;
+  if (!window.pageRenderers?.spa) {
+    return;
+  }
+  window.pageRenderers.spa.previousReactRenderers = window.pageRenderers?.spa?.reactRenderers;
   window.pageRenderers.spa.reactAppsName = [];
   window.pageRenderers.spa.reactRenderers = [];
```
Comment (cbbayburt) [inline web/html/src/core/spa/spa-renderer.ts:63]: nitpick:
```suggestion
  if (window.pageRenderers?.spa) {
    window.pageRenderers.spa.previousReactRenderers = window.pageRenderers.spa.reactRenderers;
    window.pageRenderers.spa.reactAppsName = [];
    window.pageRenderers.spa.reactRenderers = [];
  }
```

File: web/html/src/core/channels/api/use-channels-tree-api.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import { useState } from "react";
+import Network from "utils/network";
+import _keyBy from "lodash/keyBy";
+import { ChannelType } from "core/channels/type/channels.type";
+import produce from "immer";
+
+// https://github.com/facebook/flow/issues/2221
+export const getChannelsTreeValues = (channelsTree: ChannelsTreeType): Array<ChannelType> =>
+  Object.values(channelsTree.channelsById) as any;
```
Comment (cbbayburt) [inline web/html/src/core/channels/api/use-channels-tree-api.ts:9]: Bye, Flow :wave: 
```suggestion
export const getChannelsTreeValues = (channelsTree: ChannelsTreeType): Array<ChannelType> =>
  Object.values(channelsTree.channelsById);
```

File: web/html/src/core/channels/api/use-mandatory-channels-api.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+};
+
+export type UseMandatoryChannelsApiReturnType = {
+  requiredChannelsResult: RequiredChannelsResultType;
+  isDependencyDataLoaded: boolean;
+  fetchMandatoryChannelsByChannelIds: Function;
+};
+
+const useMandatoryChannelsApi = (): UseMandatoryChannelsApiReturnType => {
+  const [messages, setMessages] = useState([] as MessageType[]);
```
Comment (cbbayburt) [inline web/html/src/core/channels/api/use-mandatory-channels-api.ts:36]: Is this different than: `useState<MessageType[]>([])`?
If it's all the same, I think we should have the preference of one or the other

File: web/html/src/core/channels/utils/channels-dependencies.utils.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  requiredByChannels: Map<number, Set<number>>;
+};
+
+// Converts array of channel names into a human-readable tooltip
+// containing information about channel dependencies
+// return null if the channel is not involved in any dependencies
+function dependenciesTooltip(
+  requiredChannels: Array<string>,
+  requiredByChannels: Array<string>
+): string | null | undefined {
```
Comment (cbbayburt) [inline web/html/src/core/channels/utils/channels-dependencies.utils.ts:16]: nitpick:
```suggestion
function dependenciesTooltip(
  requiredChannels: Array<string>,
  requiredByChannels: Array<string>
): string | null {
```

**Review by cbbayburt — COMMENTED** [review]
I know we're not supposed to fix or improve anything at this stage, but I couldn't help but nitpick as I go through :see_no_evil:

File: web/html/src/core/channels/api/use-mandatory-channels-api.ts (hunk 3 above)
Comment (Etheryte) [reply web/html/src/core/channels/api/use-mandatory-channels-api.ts:36]: The resulting types are the same, I agree that the `useState<T>` pattern is better. 👍

**Comment by Etheryte** [comment]
@cbbayburt Thanks a lot for the suggestions and feedback, I've incorporated your changes and this should now be ready to merge.

**Review by cbbayburt — APPROVED** [review]
Thanks, @Etheryte. Everything looks good :+1:

**Pull request merged**